src/pool/srv_target.c

```diff
@@ -876,9 +886,25 @@ ds_pool_get(struct ds_pool *pool)
 void
 ds_pool_put(struct ds_pool *pool)
 {
-	D_ASSERT(pool != NULL);
-	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
-	daos_lru_ref_release(pool_cache, &pool->sp_entry);
+	int	rc;
+
+	/*
+	 * Someone has stopped the pool. Current user may be the one that is holding the last
+	 * reference on the pool, then drop such reference will trigger pool_free_ref() as to
+	 * stop related container that may wait current user (ULT) to exit. To avoid deadlock,
+	 * let's use independent ULT to drop the reference asynchronously and make current ULT
+	 * to go ahead.
+	 */
+	if (unlikely(pool->sp_stopping)) {
```

Comment:
It's better to introduce a daos_lru_ref_peek() to check the reference and only creating ULT when (stopping && ref == 2).

Reply:
I Has ever considered that, but it means that the caller needs to understand how the LRU reference works, for example, it is "ref == 2" instead of "ref == 1". That is too hack since it expose the LRU internal details to the  callers. Personally, I do not like such usage, because once LRU logic changed in the further, such hack may be bug. On the other hand, stopping the pool with others users referencing the pool is relative case, then even if we start another ULT, it will not cause much overhead.

Reply:
Well, if that's your concern, you could export a function for checking if it's holding the last reference.

We actually does compare vos object refcount with 2 to see if it's the last refcount. (see vos_obj_hold()), I agree it looks not that clean.

Reply:
It is not only unclean, but also seems unnecessary, because once "stopping" flag is set, nobody can find such pool (in cache) any longer. So only the former users can continuously access such pool until released the reference. So three will be at most "N - 1" (N is the system level pool users by race count, very limited) additional ULTs to be created that will neither affect the correctness nor slowdown the system too much. That is different from vos_obj_hold() case.

Reply:
After "stopping" is set, the pool can still be found until the last reference of the ds_pool being dropped and being removed from cache. See ds_pool_lookup(), the ds_pool is still in cache, so every ds_pool_lookup() will have to call ds_pool_put() to generate an ULT.

Please look into ds_pool_stop() too, it could yield and wait on many things before putting the last reference. 

What's your major concern of adding a simple function to check the last reference?

Reply:
> After "stopping" is set, the pool can still be found until the last reference of the ds_pool being dropped and being removed from cache. See ds_pool_lookup(), the ds_pool is still in cache, so every ds_pool_lookup() will have to call ds_pool_put() to generate an ULT.
 
I have changed ds_pool_lookup() logic in the patch, it will not call ds_pool_put(), instead, it will call pool_put_sync() when the found pool has "stopping" flag, that will not generate additional ULT.

> Please look into ds_pool_stop() too, it could yield and wait on many things before putting the last reference.

What's the matter for that?

> What's your major concern of adding a simple function to check the last reference?

As I mentioned above, it is not worth to introduce a hack and potential buggy "solution" but without benefit.